lib/build/builder.js

```diff
@@ -24,45 +22,36 @@ class Builder {
     }
     this.subscriptions = null
     this.goconfig = null
-    this.name = null
-    // Linter should check whether a linter is disposed, but it doesn't
-    // this.grammarScopes = null
-    this.lintOnFly = null
   }
 
-  ready () {
-    return !!this.goconfig
-  }
-
-  lint (editor) {
-    if (!this.ready() || !editor) {
-      return []
-    }
-    const p = editor.getPath()
-    if (!p) {
-      return []
+  build (editor, path = editor.getPath()) {
+    if (!isValidEditor(editor)) {
+      return Promise.reject(new Error('invalid editor'))
     }
+    const options = this.goconfig.executor.getOptions('file')
     return Promise.resolve().then(() => {
       return this.goconfig.locator.findTool('go').then((cmd) => {
         if (!cmd) {
-          return []
+          return Promise.reject(new Error('cannot find go tool'))
         }
-
-        const options = this.goconfig.executor.getOptions('file')
         const buildPromise = this.lintInstall(cmd, options)
-
-        const testPromise = this.hasTests(p)
+        const testPromise = this.hasTests(path)
           ? this.lintTest(cmd, options)
-          : Promise.resolve({output: '', linterName: 'test'})
+          : Promise.resolve({output: '', linterName: 'test', exitcode: 0})
 
-        return Promise.all([buildPromise, testPromise]).then((results) => this.getMessages(results, options.cwd))
+        return Promise.all([buildPromise, testPromise])
       })
-    }).catch((error) => {
-      if (error.handle) {
-        error.handle()
+    }).then((results) => {
+      if (results && results.length) {
+        const messages = this.getMessages(results, options.cwd)
+        this.linter.setMessages(messages)
```

Comment:
Set messages is only being invoked once here?

This is one thing v2 linter does better is distinguishes between set messages and set all messages.